### DISCUSSION TITLE
Missing VertexFormat::Unorm8x3 and VertexFormat::Unorm16x3

### DISCUSSION THREAD

**Discussion by Pauan** [gfx-rs/wgpu#1769] [Q&A] [answered]
Currently there are `x2` and `x4` variants of [`Unorm`](https://docs.rs/wgpu/0.9.0/wgpu/enum.VertexFormat.html), but not `x3`. Is this intentional or is it just a missing feature?

How am i supposed to send RGB data to the GPU (which is `[u8; 3]`)? Do I have to manually pad every array to make it into a `[u8; 4]`?

The glTF parser crate supports [both RGB and RGBA in various formats](https://docs.rs/gltf/0.16.0/gltf/mesh/util/enum.ReadColors.html), and ideally I would like to support all of them.

**Comment by kvark** [comment, marked as answer]
These vertex formats are not portable, hence missing from WebGPU upstream and `wgpu`.
What is the use case for providing lots of RGB data to vertex attributes?

**Reply by Pauan** [reply]
Well, that's what vertex colors are. Though I'm not sure how often glTF models use RGB vertex colors rather than RGBA vertex colors, so perhaps it's a non-issue in practice.

**Reply by cwfitzgerald** [reply]
In general I'd recommend normalizing things like vertex formats to prevent combinatoric explosions of the pipeline states needed to deal with gltf.

**Reply by Pauan** [reply]
@cwfitzgerald Yeah, that makes sense, though WGSL already does a lot of normalization.

For example it will auto-convert a `vec3` into a `vec4` (adding `1` to the final element). And the point of `Unorm` is to normalize an integer into an `f32`, so the only thing I'm missing is `Unorm8x3` and `Unorm16x3`.

But since they're never going to be supported, I guess I have to do manual normalization before passing to the buffer. Unfortunately that does have a pretty big performance cost, since I can't just pass the bytes directly into the buffer, I'm forced to allocate a new `Vec<u8>`.
